services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/validation/HankeValidator.kt

```diff
@@ -19,92 +27,41 @@ class HankeValidator : ConstraintValidator<ValidHanke, Hanke> {
             return false
         }
 
-        var ok = true
-        if (hanke.nimi.isNullOrBlank()) {
-            context
-                .buildConstraintViolationWithTemplate(HankeError.HAI1002.toString())
-                .addPropertyNode("nimi")
-                .addConstraintViolation()
-            ok = false
-        }
-
-        if (hanke.vaihe == null) {
-            context
-                .buildConstraintViolationWithTemplate(HankeError.HAI1002.toString())
-                .addPropertyNode("vaihe")
-                .addConstraintViolation()
-            ok = false
-        } else if (hanke.vaihe!! == Vaihe.SUUNNITTELU && hanke.suunnitteluVaihe == null) {
-            // if vaihe = SUUNNITTELU then suunnitteluVaihe must have value
-            context
-                .buildConstraintViolationWithTemplate(HankeError.HAI1002.toString())
-                .addPropertyNode("suunnitteluVaihe")
-                .addConstraintViolation()
-            ok = false
-        }
+        val hankeResult = hanke.validate()
+        hankeResult.errorPaths().forEach { context.addViolation(HankeError.HAI1002, it) }
 
-        ok = ok && checkHankealueet(hanke, context)
-        ok = ok && checkTyomaaTiedot(hanke, context)
+        val alueResult = validate().andAllIn(hanke.alueet, "alueet", ::validateHankeAlue)
+        alueResult.errorPaths().forEach { context.addViolation(HankeError.HAI1032, it) }
 
-        return ok
+        return hankeResult.isOk() and alueResult.isOk()
```

Comment:
Nitpicking but doesn't 'and' always evaluate both sides? If hankeResult is not ok, is it necessary to check alueresult.ok?

Reply:
Sorry, it doesn't actually matter here. I think I had the `and` between the validation calls in an earlier iteration. Now the errors are already added to the context and this just checks if both booleans are true.

Reply:
Changed to `&&`.